lib/hotel.rb

```diff
@@ -0,0 +1,168 @@
+  end
+
+  def self.find_reservation_by_block_id(input_id)
+    raise ArgumentError.new "Invalid input.  Please enter Block ID as an Integer" if !(input_id.is_a? Integer)
+    blocked_reservations = []
+    all_reservations = self.all_reservations
+    all_reservations.each do |reservation|
+      blocked_reservations << reservation if reservation.block_id == input_id
+    end
+    raise ArgumentError.new "No Reservations found under that Block ID" if blocked_reservations.length == 0
+    return blocked_reservations
+  end
+
+  def self.find_block(input_id)
+    all_blocks = self.all_blocks
+    all_blocks.each do |block|
+      return block if block.block_id == input_id
+    end
+  end
+
+  def self.all_reservations
+    all_reservations = []
+    CSV.read('support/reservations.csv').each do |row|
+      reservation_id = row[0]
+      room_num = row[1]
+      check_in_date = row[2].split
+      check_out_date = row[3].split
+      block_id = row[4] ? row[4] : 0
+      all_reservations.push(Hotel::Reservation.new(reservation_id, room_num, check_in_date, check_out_date, block_id))
+    end
+    return all_reservations
+  end
+
+  def self.all_blocks
+    all_blocks =[]
+    CSV.read('support/blocks.csv').each do |row|
+      block_rate = row[0]
+      num_of_rooms = row[1]
+      check_in_date = row[2].split
+      check_out_date = row[3].split
+      block_id = row[4]
+      all_blocks.push(Hotel::Block.new(block_rate, num_of_rooms, check_in_date, check_out_date, block_id))
+    end
+    return all_blocks
+  end
+
+  def self.cost(input_reservation_id)
+    raise ArgumentError.new "Invalid reservation ID.  Must be Integer." if !(input_reservation_id.is_a? Integer)
+    all_reservations = self.all_reservations
+    all_reservations.each do |reservation|
+      return (reservation.total_cost) if reservation.id == input_reservation_id
+    end
+    raise ArgumentError.new "Reservation ID does not exist"
+  end
+
+  def self.access_reservation(input_date)
+    search_date = Date.new(input_date[0], input_date[1], input_date[2])
+    search_reservations = []
+    all_reservations = self.all_reservations
+    all_reservations.each do |reservation|
+      search_reservations << reservation if search_date.between?(reservation.check_in, reservation.check_out - 1)
+    end
+    raise ArgumentError.new "No Reservations are present on that date" if search_reservations.length == 0
+    return search_reservations
+  end
+
+  def self.available_rooms(begin_date, end_date)
+    begin_search = Date.new(begin_date[0].to_i, begin_date[1].to_i, begin_date[2].to_i)
+    end_search = Date.new(end_date[0].to_i, end_date[1].to_i, end_date[2].to_i)
+    unavailable_rooms = []
+    all_reservations = self.all_reservations
+    all_reservations.each do |reservation|
+      if (begin_search >= reservation.check_in) && (begin_search < reservation.check_out) && (end_search >= reservation.check_in) && (end_search <= reservation.check_out)
```

Comment:
maybe this logic can be returned from a reservation itself if you pass in the begin_search and end_search properties instead! That way, hotel doesn't need to know that a reservation has a `.check_in` and a `.check_out`